include/pops/generator_provider.hpp

```diff
@@ -0,0 +1,515 @@
+{
+public:
+    /**
+     * @brief Seeds the underlying generator
+     * @param seed for the underlying generator
+     */
+    SingleGeneratorProvider(unsigned seed)
+    {
+        this->seed(seed);
+    }
+
+    /* Re-seed the generator */
+    void seed(unsigned seed)
+    {
+        general_generator_.seed(seed);
+    }
+
+    /* This overload always throws std::invalid_argument because only one seed is
+     * supported. */
+    void seed(const std::map<std::string, unsigned>& seeds)
+    {
+        UNUSED(seeds);
+        throw std::invalid_argument(
+            "Multiple seeds are not supported by SimpleGeneratorProvider (only one seed is supported)");
+    }
+
+    /* Re-seeds the generator
+     *
+     * Throws std::invalid_argument if configuration contains multiple seeds.
+     */
+    void seed(const Config& config)
+    {
+        if (config.multiple_random_seeds) {
+            throw std::invalid_argument(
+                "Config cannot have multiple_random_seeds set for SimpleGeneratorProvider (only random_seed is supported)");
+        }
+        if (!config.random_seeds.empty()) {
+            throw std::invalid_argument(
+                "Config cannot have random_seeds set for SimpleGeneratorProvider (only random_seed is supported)");
+        }
+        general_generator_.seed(config.random_seed);
+    }
+
+    Generator& general()
+    {
+        return general_generator_;
+    }
+
+    Generator& disperser_generation()
+    {
+        return general();
+    }
+
+    Generator& natural_dispersal()
+    {
+        return general();
+    }
+
+    Generator& anthropogenic_dispersal()
+    {
+        return general();
+    }
+
+    Generator& establishment()
+    {
+        return general();
+    }
+
+    Generator& weather()
+    {
+        return general();
+    }
+
+    Generator& movement()
+    {
+        return general();
+    }
+
+    Generator& overpopulation()
+    {
+        return general();
+    }
+
+    Generator& survival_rate()
+    {
+        return general();
+    }
+
+    Generator& soil()
+    {
+        return general();
+    }
+
+    // API to behave like the underlying generator.
+
+    using result_type = typename Generator::result_type;
+
+    static result_type min()
+    {
+        return Generator::min();
+    }
+
+    static result_type max()
+    {
+        return Generator::max();
+    }
+
+    result_type operator()()
+    {
+        return general_generator_();
+    }
+
+    void discard(unsigned long long n)
+    {
+        general_generator_.disard(n);
+    }
+
+private:
+    Generator general_generator_;
+};
+
+/** Default generator provider to be used for tests and development */
+using DefaultSingleGeneratorProvider =
+    SingleGeneratorProvider<std::default_random_engine>;
+
+/** Generator provider providing multiple isolated generators
+ *
+ * All ways of seeding result is multiple independent generators.
+ */
+template<typename Generator>
+class IsolatedRandomNumberGeneratorProvider
```

Comment:
Consider renaming to MultiRandom...

Reply:
Renamed to Multi... together with variables named "isolated". I also made the error messages more consistent with the documentation using "multiple isolated generators".